compiler/qsc_frontend/src/typeck/tests.rs

```diff
@@ -4415,3 +4440,236 @@ fn within_block_should_be_unit_error() {
         "##]],
     );
 }
+
+#[test]
+fn path_field_access() {
+    check(
+        indoc! {"
+            namespace A {
+                struct B { C : Int }
+                function Foo() : Unit {
+                    let b = new B { C = 5 };
+                    b.C;
+                }
+            }
+        "},
+        "",
+        &expect![[r##"
+            #14 55-57 "()" : Unit
+            #18 65-118 "{\n        let b = new B { C = 5 };\n        b.C;\n    }" : Unit
+            #20 79-80 "b" : UDT<"B": Item 1>
+            #22 83-98 "new B { C = 5 }" : UDT<"B": Item 1>
+            #27 95-96 "5" : Int
+            #29 108-111 "b.C" : Int
+            #31 108-109 "b" : UDT<"B": Item 1>
+            #32 110-111 "C" : Int
+        "##]],
+    );
+}
+
+#[test]
+fn field_access_chained() {
+    check(
+        indoc! {"
+            namespace A {
+                struct B { C : Int }
+                struct D { E : B }
+                function Foo() : Unit {
+                    let d = new D { E = new B { C = 5 } };
+                    d.E.C;
+                }
+            }
+        "},
+        "",
+        &expect![[r##"
+            #22 78-80 "()" : Unit
+            #26 88-157 "{\n        let d = new D { E = new B { C = 5 } };\n        d.E.C;\n    }" : Unit
+            #28 102-103 "d" : UDT<"D": Item 2>
+            #30 106-135 "new D { E = new B { C = 5 } }" : UDT<"D": Item 2>
+            #35 118-133 "new B { C = 5 }" : UDT<"B": Item 1>
+            #40 130-131 "5" : Int
+            #42 145-150 "d.E.C" : Int
+            #44 145-146 "d" : UDT<"D": Item 2>
+            #45 147-148 "E" : UDT<"B": Item 1>
+            #46 149-150 "C" : Int
+        "##]],
+    );
+}
+
+#[test]
+fn expr_field_access() {
+    check(
+        indoc! {"
+            namespace A {
+                struct B { C : Int }
+                function Foo() : Unit {
+                    (new B { C = 5 }).C;
+                }
+            }
+        "},
+        "",
+        &expect![[r##"
+            #14 55-57 "()" : Unit
+            #18 65-101 "{\n        (new B { C = 5 }).C;\n    }" : Unit
+            #20 75-94 "(new B { C = 5 }).C" : Int
+            #21 75-92 "(new B { C = 5 })" : UDT<"B": Item 1>
+            #22 76-91 "new B { C = 5 }" : UDT<"B": Item 1>
+            #27 88-89 "5" : Int
+        "##]],
+    );
+}
+
+#[test]
+fn expr_incomplete_field_access() {
+    check_allow_parse_errors(
+        indoc! {"
+            namespace A {
+                struct B { C : Int }
+                function Foo() : Unit {
+                    (new B { C = 5 }).;
```

Comment:
What is the behavior when you don't end this line in a semicolon? Does it complain that the expected return type of `Unit` doesn't match `?`?

Reply:
Good question, I should add a test case for that since that would emulate real-life typing more closely. What I'm looking for is for the type of the `(new B { C = 5 })` to be inferred correctly, because that will help us gather the field names for completions.

Though these tests don't validate typecheck errors, I can see from the output: Yes, the overall type of this block does become `?` to match the type of the broken expression. Which doesn't match `Unit`.